File: android/app/src/main/java/com/testlogon/android/feature/markets/InstrumentClass.kt
package com.testlogon.android.feature.markets

import com.testlogon.android.data.exchange.Instrument

/**
 * The instrument classes the unified, class-filtered symbol picker groups markets by. Built PURELY
 * from confirmed backend discriminators (see [classesForSymbol]); no invented fields:
 *  - [SPOT]: a non-perpetual (cash) instrument (`isPerpetual == false`).
 *  - [PERP]: a perpetual future (`isPerpetual == true`).
 *  - [PREDICTION]: a symbol with a live binary prediction-market state (detected by probing the
 *    per-symbol PM-state read; there is no list endpoint).
 *  - [FUNDING]: a lens on perpetuals with a positive funding interval — surfaces the funding
 *    interval + latest applied funding rate (bps).
 *
 * A single symbol can belong to several classes at once (a perp is also a FUNDING-book row; a
 * PM-enabled symbol adds PREDICTION), which is why [classesForSymbol] returns a set.
 */
enum class InstrumentClass(val label: String) {
    SPOT("Spot"),
    PERP("Perp"),
    PREDICTION("Prediction"),
    FUNDING("Funding"),
}

/**
 * A filter tab in the picker: either [InstrumentClass.All] (everything) or one concrete class. Kept
 * as a small sealed hierarchy so the "All" tab is a first-class, exhaustive option rather than a
 * nullable class.
 */
sealed interface MarketClassTab {
    val label: String

    data object All : MarketClassTab {
        override val label: String = "All"
    }

    data class Of(val clazz: InstrumentClass) : MarketClassTab {
        override val label: String = clazz.label
    }
}

/** The ordered tab row shown across the top of the picker: All, then Spot, Perp, Prediction, Funding. */
val MARKET_CLASS_TABS: List<MarketClassTab> = buildList {
    add(MarketClassTab.All)
    // Preserve the enum declaration order for the concrete tabs.
    InstrumentClass.entries.forEach { add(MarketClassTab.Of(it)) }
}

/**
 * PURE classifier: the set of [InstrumentClass]es a [symbol] belongs to.
 *
 * @param isPrediction whether a live PM state was observed for this symbol (probed elsewhere; treat
 *   404/error/unknown as false so a non-PM symbol never claims [InstrumentClass.PREDICTION]).
 *
 * Rules (all from confirmed discriminators):
 *  - exactly one of SPOT / PERP by `isPerpetual`.
 *  - a perp with `fundingIntervalS > 0` is ALSO a FUNDING-book row.
 *  - `isPrediction` adds PREDICTION on top of its spot/perp base class.
 */
fun classesForSymbol(symbol: Instrument, isPrediction: Boolean): Set<InstrumentClass> {
    val classes = LinkedHashSet<InstrumentClass>()
    if (symbol.isPerpetual) {
        classes.add(InstrumentClass.PERP)
        if (symbol.fundingIntervalS > 0) classes.add(InstrumentClass.FUNDING)
    } else {
        classes.add(InstrumentClass.SPOT)
    }
    if (isPrediction) classes.add(InstrumentClass.PREDICTION)
    return classes
}

/**
 * Whether a symbol (with its probed [isPrediction] flag) belongs under the given [tab]. "All" always
 * matches; a concrete tab matches when the symbol's class set contains that class.
 */
fun matchesTab(symbol: Instrument, isPrediction: Boolean, tab: MarketClassTab): Boolean = when (tab) {
    is MarketClassTab.All -> true
    is MarketClassTab.Of -> classesForSymbol(symbol, isPrediction).contains(tab.clazz)
}

File: frontend/src/lib/instrumentClass.ts
/**
 * Pure instrument classifier — the single source of truth for which "class"
 * a market symbol belongs to, shared by the Markets list and the Cmd+K palette
 * so they behave as ONE logical symbol picker.
 *
 * A symbol can belong to SEVERAL classes at once:
 *   - a perpetual is BOTH a "perp" AND (when funding_interval_s > 0) a "funding" book entry;
 *   - a symbol with a live prediction-market state ADDS "prediction".
 * Discriminators are the CONFIRMED backend fields only — do NOT invent new ones:
 *   - spot        := is_perpetual === false
 *   - perp        := is_perpetual === true
 *   - funding     := is_perpetual === true && funding_interval_s > 0 (a lens on perps)
 *   - prediction  := has a live PM state (detected out-of-band; passed in as a flag)
 */
import type { MarketSymbol } from "@/api/endpoints/marketData";

export type InstrumentClass = "spot" | "perp" | "prediction" | "funding";

/** The "All" pseudo-tab value used by the filter UI (not a real class). */
export const ALL_CLASS = "all" as const;
export type ClassTab = typeof ALL_CLASS | InstrumentClass;

/** Human labels for each class + the All pseudo-tab. */
export const CLASS_LABELS: Record<ClassTab, string> = {
  all: "All",
  spot: "Spot",
  perp: "Perp",
  prediction: "Prediction",
  funding: "Funding",
};

/** Display order for the class tabs / palette headings (All first). */
export const CLASS_TAB_ORDER: ClassTab[] = ["all", "spot", "perp", "prediction", "funding"];

/** Display order of the real classes (no All) — used for palette headings. */
export const CLASS_ORDER: InstrumentClass[] = ["spot", "perp", "prediction", "funding"];

/** Honest empty-state copy per class (shown when a filtered group is empty). */
export const CLASS_EMPTY_COPY: Record<InstrumentClass, string> = {
  spot: "No spot markets listed.",
  perp: "No perpetual contracts listed.",
  prediction: "No prediction markets listed.",
  funding: "No funding-book contracts listed.",
};

export interface ClassifyOpts {
  /** True when the symbol has a live prediction-market state (probed out-of-band). */
  isPrediction?: boolean;
}

/**
 * All classes a symbol belongs to, in {@link CLASS_ORDER}. A perp is also in
 * the funding book; a PM-enabled symbol additionally gets "prediction".
 */
export function classesForSymbol(sym: MarketSymbol, opts: ClassifyOpts = {}): InstrumentClass[] {
  const out: InstrumentClass[] = [];
  if (sym.is_perpetual) {
    out.push("perp");
    if ((sym.funding_interval_s ?? 0) > 0) out.push("funding");
  } else {
    out.push("spot");
  }
  if (opts.isPrediction) out.push("prediction");
  // Normalize to the canonical display order.
  return CLASS_ORDER.filter((c) => out.includes(c));
}

/** True when the symbol belongs to the given class tab ("all" always matches). */
export function symbolInClass(
  sym: MarketSymbol,
  tab: ClassTab,
  opts: ClassifyOpts = {},
): boolean {
  if (tab === ALL_CLASS) return true;
  return classesForSymbol(sym, opts).includes(tab);
}
